File: example/services/json_placeholder_service.dart
import 'package:service_client/service_client.dart';

class JsonPlaceholderService {
  static final ServiceClientConfig _config = ServiceClientConfig(
    baseUrl: Uri.parse('https://jsonplaceholder.typicode.com'),
    defaultHeaders: {
      'Content-Type': 'application/json',
    },
    timeout: const Duration(seconds: 30),
    auth: false, // This API does not use token-based auth
  );

  /// Lazy initialization of the ServiceClient instance.
  static ServiceClient? _client;
  static ServiceClient get _service {
    _client ??= HttpServiceClient(_config);
    return _client!;
  }

  /// Fetches a todo item by its ID.
  static Future<Map<String, dynamic>> getTodo(int id) async {
    final request = ServiceRequest.http(
      method: 'GET',
      endpoint: 'todos/$id',
      errorMessage: 'Failed to fetch TODO from JSONPlaceholder',
    );

    final response = await _service.send(
      request,
    );

    if (response.statusCode >= 200 && response.statusCode < 300) {
      return response.data as Map<String, dynamic>;
    }

    throw Exception(
      'JSONPlaceholder error: ${response.statusCode} - ${response.rawBody}',
    );
  }
}

File: code/ts/example/services/json_placeholder_service.ts
import {
  type Result,
  success,
  failure,
  ServiceClientConfig,
  ServiceRequest,
  type ServiceClient,
  HttpServiceClient,
  HttpClientException,
} from '../../src';
import { ToDo } from '../models/todo';
import { ToDoFailure } from '../models/todo_failure';

/// Servicio que conecta con la API pública de JSONPlaceholder.
export class JsonPlaceholderService {
  private static readonly config = new ServiceClientConfig({
    baseUrl: 'https://jsonplaceholder.typicode.com',
    defaultHeaders: { 'Content-Type': 'application/json' },
    timeout: 30_000,
  });

  private static client: ServiceClient | undefined;
  private static get service(): ServiceClient {
    this.client ??= new HttpServiceClient(this.config);
    return this.client;
  }

  /// Obtiene un ToDo por su id.
  ///
  /// Retorna Success<ToDo> si el request fue exitoso,
  /// o Failure<ToDoFailure> si hubo un error HTTP.
  static async getTodo(id: number): Promise<Result<ToDo, ToDoFailure>> {
    const request = ServiceRequest.http({
      method: 'GET',
      endpoint: `todos/${id}`,
      errorMessage: 'Failed to fetch TODO from JSONPlaceholder',
    });

    try {
      const response = await this.service.send(request);
      const data = response.data as Record<string, unknown>;
      return success(ToDo.fromJson(data));
    } catch (e) {
      if (e instanceof HttpClientException) {
        return failure(
          new ToDoFailure({
            statusCode: e.statusCode,
            message: e.message,
            responseBody: e.response,
          }),
        );
      }
      throw e;
    }
  }
}
